main.cc

```diff
@@ -404,10 +440,15 @@ class Peer : public SocketHandler {
 		int r = connect(sd, (struct sockaddr *)&addr_, sizeof(addr_));
 
 		stat.tcp_handshakes++;
+		state_ = STATE_TCP_CONNECTING;
 
 		// On on localhost connect() can complete instantly
-		// even on non-blocking sockets.
-		return tcp_connect_try_finish(r);
+		// even on non-blocking sockets (Tempesta FW case).
```

Comment:
Not only in Tempesta FW case but also in cases when host is unreachable (no route or connection refused). This apply to comment only, code the works good.